libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -1,221 +1,513 @@
+}
+
+// Allow unit tests to supply a dummy StartProxy.
+var startProxy = portmapper.StartProxy
+
+// attemptBindHostPorts allocates host ports for each port mapping that requires
+// one, and reserves those ports by starting docker-proxy.
+//
+// If the allocator doesn't have an available port in the required range, or the
+// docker-proxy process doesn't start (perhaps because another process has
+// already bound the port), all resources are released and an error is returned.
+// When ports are successfully reserved, a portBinding is returned for each
+// mapping.
+//
+// If NAT is disabled for any of the bindings, no host port reservation is
+// needed. These bindings are included in results, as the container port itself
+// needs to be opened in the firewall.
+func attemptBindHostPorts(
+	cfg []*portBindingReq,
+	proto string,
+	hostPortStart, hostPortEnd uint16,
+	proxyPath string,
+) (_ []*portBinding, retErr error) {
+	var err error
+	var port int
+
+	addrs := make([]net.IP, 0, len(cfg))
+	for _, c := range cfg {
+		if !c.disableNAT {
+			addrs = append(addrs, c.HostIP)
+		}
+	}
+	if len(addrs) > 0 {
+		pa := portallocator.Get()
+		port, err = pa.RequestPortsInRange(addrs, proto, int(hostPortStart), int(hostPortEnd))
+		if err != nil {
+			return nil, err
+		}
+		defer func() {
+			if retErr != nil {
+				for _, a := range addrs {
+					pa.ReleasePort(a, proto, port)
+				}
+			}
+		}()
 	}
 
-	// Save the host port (regardless it was or not specified in the binding)
-	switch netAddr := host.(type) {
-	case *net.TCPAddr:
-		bnd.HostPort = uint16(host.(*net.TCPAddr).Port)
-		return nil
-	case *net.UDPAddr:
-		bnd.HostPort = uint16(host.(*net.UDPAddr).Port)
-		return nil
-	case *sctp.SCTPAddr:
-		bnd.HostPort = uint16(host.(*sctp.SCTPAddr).Port)
-		return nil
-	default:
-		// For completeness
-		return ErrUnsupportedAddressType(fmt.Sprintf("%T", netAddr))
+	res := make([]*portBinding, 0, len(cfg))
+	for _, c := range cfg {
+		cc := c.PortBinding.GetCopy()
+		pb := &portBinding{PortBinding: &cc}
+		if c.disableNAT {
+			pb.HostPort = 0
+		} else {
+			pb.proxy, err = startProxy(c.Proto.String(), c.HostIP, port, c.IP, int(c.Port), proxyPath)
+			if err != nil {
+				return nil, fmt.Errorf("failed to bind port %s:%d/%s: %w", c.HostIP, port, c.Proto, err)
+			}
+			defer func() {
+				if retErr != nil {
+					pb.proxy.Stop()
+				}
+			}()
+			pb.HostPort = uint16(port)
+		}
+		pb.HostPortEnd = pb.HostPort
+		res = append(res, pb)
 	}
+	return res, nil
 }
 
+// releasePorts attempts to release all port bindings, does not stop on failure
 func (n *bridgeNetwork) releasePorts(ep *bridgeEndpoint) error {
-	return n.releasePortsInternal(ep.portMapping)
-}
+	n.Lock()
+	pm := ep.portMapping
+	ep.portMapping = nil
+	n.Unlock()
 
-func (n *bridgeNetwork) releasePortsInternal(bindings []types.PortBinding) error {
-	var errorBuf bytes.Buffer
+	var err error
+	for _, b := range pm {
+		err = errors.Join(err, n.releasePortBinding(b))
+	}
+	return err
```

Comment:
Done - by merging with `releasePortBinding` to make `releasePortBindings`, because `releasePortBinding` was called elsewhere but also with a list of `portBinding`.